fireworks/core/mongo_launchpad.py

```diff
@@ -0,0 +1,1407 @@
+from fireworks.utilities.fw_serializers import recursive_dict
+
+
+__author__ = 'Anubhav Jain'
+__copyright__ = 'Copyright 2013, The Materials Project'
+__version__ = '0.1'
+__maintainer__ = 'Anubhav Jain'
+__email__ = 'ajain@lbl.gov'
+__date__ = 'Jan 30, 2013'
+
+# TODO: lots of duplication reduction and cleanup possible
+
+class WFLock(object):
+    """
+    Lock a Workflow, i.e. for performing update operations
+    Raises a LockedWorkflowError if the lock couldn't be acquired withing expire_secs and kill==False.
+    Calling functions are responsible for handling the error in order to avoid database inconsistencies.
+    """
+
+    def __init__(self, lp, fw_id, expire_secs=WFLOCK_EXPIRATION_SECS, kill=WFLOCK_EXPIRATION_KILL):
+        """
+        Args:
+            lp (LaunchPad)
+            fw_id (int): Firework id
+            expire_secs (int): max waiting time in seconds.
+            kill (bool): force lock acquisition or not
+        """
+        self.lp = lp
+        self.fw_id = fw_id
+        self.expire_secs = expire_secs
+        self.kill = kill
+
+    def __enter__(self):
+        ctr = 0
+        waiting_time = 0
+        # acquire lock
+        links_dict = self.lp.workflows.find_one_and_update({'nodes': self.fw_id,
+                                                            'locked': {"$exists": False}},
+                                                           {'$set': {'locked': True}})
+        # could not acquire lock b/c WF is already locked for writing
+        while not links_dict:
+            ctr += 1
+            time_incr = ctr/10.0+random.random()/100.0
+            time.sleep(time_incr)  # wait a bit for lock to free up
+            waiting_time += time_incr
+            if waiting_time > self.expire_secs:  # too much time waiting, expire lock
+                wf = self.lp.workflows.find_one({'nodes': self.fw_id})
+                if not wf:
+                    raise ValueError("Could not find workflow in database: {}".format(self.fw_id))
+                if self.kill:  # force lock acquisition
+                    self.lp.m_logger.warning('FORCIBLY ACQUIRING LOCK, WF: {}'.format(self.fw_id))
+                    links_dict = self.lp.workflows.find_one_and_update({'nodes': self.fw_id},
+                                                                       {'$set': {'locked': True}})
+                else:  # throw error if we don't want to force lock acquisition
+                    raise LockedWorkflowError("Could not get workflow - LOCKED: {}".format(self.fw_id))
+            else:
+                # retry lock
+                links_dict = self.lp.workflows.find_one_and_update(
+                    {'nodes': self.fw_id, 'locked': {"$exists": False}}, {'$set': {'locked': True}})
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.lp.workflows.find_one_and_update({"nodes": self.fw_id}, {"$unset": {"locked": True}})
+
+
+class MongoLaunchPad(LaunchPad, FWSerializable):
```

Comment:
LaunchPad is already FWSerializable, so it shouldn't be here again.

Reply:
Fixed in 5ddbea4e8f4696b2c20bf453397932146d7a6536